sql/src/main/java/io/crate/operation/projectors/BatchIteratorBackpressureExecutor.java

```diff
@@ -55,145 +54,120 @@
     private final Function<Boolean, CompletableFuture<R>> executeFunction;
     private final Consumer<Row> onRowConsumer;
     private final ScheduledExecutorService scheduler;
-    private final AtomicBoolean collectingEnabled = new AtomicBoolean(false);
-    private final AtomicBoolean isScheduledCollectionRunning = new AtomicBoolean(false);
-    private final AtomicBoolean computeFinalResult = new AtomicBoolean(false);
-    private final Iterator<TimeValue> consumeIteratorDelays;
-    private final BooleanSupplier backPressureTrigger;
-    private final BiConsumer<R, Throwable> bulkExecutionCompleteListener;
-    private final BiConsumer<Object, Throwable> loadNextBatchCompleteListener;
-    private final Runnable scheduleConsumeIteratorJob;
+    private final Iterator<TimeValue> throttleDelay;
+    private final BooleanSupplier pauseConsumption;
+    private final BiConsumer<Object, Throwable> continueConsumptionOrFinish;
     private final int bulkSize;
     private final AtomicInteger inFlightExecutions = new AtomicInteger(0);
     private final CompletableFuture<Void> resultFuture = new CompletableFuture<>();
+    private final Semaphore semaphore = new Semaphore(1);
 
-    private int indexInBulk = 0;
-    private volatile boolean lastBulkScheduledToExecute = false;
+    private int indexInBulk = -1;
```

Comment:
wouldn't this cause us to have an extra item in the first bulk? (ie if the bulkSize is 10, we'll first send 11 items)